Fortify db and log-db gen-servers

owaddell opened the pull request:
**Fixes**

* Let applications selectively tolerate `log-handler` faults.
* Make `db` more robust in shutdown and in its handling of transactions.
* Expose options for tuning database parameters.
* Document lazy transactions more accurately.

**Proposed changes**

Summarize changes here. Note any breaking changes.

* Recast `define-options` using native records in place of parameter-setting.
* Add compile-time string concatenation utilities.
* Add `make-swish-event-logger` to permit control over `prune-max-days` and `prune-limit` for the Swish event tables.
* Expose `create-prune-on-insert-trigger`.
* Expose `commit-limit`, `commit-delay`, and `cache-timeout` database parameters.
* Add `db:expire-cache` for explicit control over statement cache.
* Allow `db:options` argument to `log-db` to permit tuning the log database.
* Add `log-db:event-logger` interface so we can extend logger interface without breaking `<event-logger>`.
* Add optional `endure?` predicate to `event-mgr:set-log-handler` so applications have the option of tolerating log-handler faults, without paying the extra `try` / `catch` overhead.
* Add optional `tolerate-fault` (hmm, should that be `tolerate-fault?`) to `log-db:event-logger` to take advantage of `event-mgr:set-log-handler`'s `endure?` option.
* Make `db` gen-server `flush` its work queue on `shutdown`.
* Handle failure while finalizing statements when database is busy at `shutdown`.
* Overhaul `db` gen-server to use `monitor` in place of `link` to make it more robust to misbehaving transactions.
* Adapt the `supervisor` shutdown protocol in `db` gen-server's `terminate` path.

owaddell-beckman committed:
remove unreferenced let binding

owaddell-beckman committed:
add commit-limit to new db:options

owaddell-beckman committed:
test: cover case where log-handler is already set

owaddell-beckman committed:
add ct:join and ct:string-append

owaddell-beckman committed:
make valid-fields? available for other macros

- valid-fields? is useful for other (internal) macros
- more precise syntax error for bad tuple fields

owaddell-beckman committed:
test: add clarifying example to native-record mat

Forgetting the (nongenerative) clause creates confusion since
match may work with a top-level define-record-type, but fail
when the same definition is local. Adding (nongenerative) is
typically the fix.

owaddell-beckman committed:
revise define-options

The original implementation was elegant, but it relied on parameters
and had to be used with care.

The new implementation simply builds native records. It supports many
features of define-tuple, such as copy, copy*, and field accessors, yet
adds support for optional fields, default values, predicates, and
filters. The aim is to allow us to extend configurable interfaces
without treating each extension as a breaking change.

owaddell-beckman committed:
add cache-timeout to db:options

payneca committed:
add db:expire-cache

owaddell-beckman committed:
add commit-delay to db:options

payneca committed:
improve database gen-server's fault tolerance

terminate
* calls update before flush to process messages even if a worker is
  not currently running.
* allow a failure while finalizing statements and closing the
  database. In practice this was a UV_EBUSY, and we're assuming that
  if one fails the following calls will also fail for the same reason.

update
* changed the design to use monitors instead of links allowing
  robustness against transactions where user-code unlinks from the
  database.

flush
* processes the entire work queue until failure.
* interrupts the C thread with sqlite:interrupt reducing the
  possibility of UV_EBUSY within terminate.
* employs supervisor shutdown protocol if necessary.

owaddell-beckman committed:
add endure? option to event-mgr:set-log-handler

Let applications flexibly respond to log-handler faults while taking
advantage of the try/catch already in place where log-handler is called.

owaddell-beckman committed:
expose create-prune-on-insert-trigger

owaddell-beckman committed:
add make-swish-event-logger

Let applications tune prune triggers for the Swish event tables.

owaddell-beckman committed:
add log-db:event-logger tolerate-fault?

Let applications configure fault handling for log procedure
without incurring overhead of added try/catch.

payneca committed:
clarify release and development version numbers

owaddell-beckman committed:
update osi_interrupt_database and sqlite:interrupt to return the database busy state

owaddell-beckman committed:
doc: describe lazy transactions more accurately

owaddell-beckman committed:
doc: move text closer to use of constants

owaddell-beckman committed:
test: make sure data-dir exists

owaddell-beckman committed:
allow db:options argument to log-db:start&link

Allow applications to tune database parameters for the log-db.

burgerrg commented on doc/swish/db.tex:
```diff
@@ -206,14 +213,20 @@
-\item \code{`(EXIT \var{worker-pid} normal)}: The worker finished
-  the previous request successfully. Process the queue.
+\item \code{`(DOWN \_ \_ \_)}: An unexpected \code{DOWN} message is
+  ignored and the server processes the queue.
```
Suggestion: `Ignore the unexpected DOWN message. Process the queue.`

laqrix replied:
Updating as suggested.

burgerrg commented on doc/swish/db.tex:
```diff
@@ -206,14 +213,20 @@
-  failed to process the previous request. Flush the queue and stop
-  with \var{reason}.
+\item \code{`(EXIT \var{pid} \_ \var{e})}: If the \var{pid} is the
+  worker, ignore the message. Do not update the state. A followup
```
`followup` => `follow-up`

laqrix replied:
Updating as suggested.

burgerrg commented on doc/swish/db.tex:
```diff
@@ -206,14 +213,20 @@
+\item \code{`(EXIT \var{pid} \_ \var{e})}: If the \var{pid} is the
+  worker, ignore the message. Do not update the state. A followup
+  \code{DOWN} message will process the queue. Otherwise, flush the
+  queue and stop with the fault \var{e}.
```
Add a comma after `queue`.

laqrix replied:
Updating as suggested.

owaddell-beckman committed:
add define-tuple (type is?)

Support (type is?) that expands into (lambda (x) (type is? x))
just as (type field) expands into (lambda (x) (type field x)).

owaddell-beckman committed:
add log-db:event-logger

burgerrg commented on doc/swish/erlang.tex:
```diff
@@ -1656,6 +1656,15 @@ \subsection{Miscellaneous}
+\begin{syntax}
+  \code{(\var{name} is?)}
+\end{syntax}
+\expandsto{} a predicate that returns true if its argument is an instance
```
`if` => `if and only if`

laqrix replied:
Updating as suggested.

burgerrg commented on doc/swish/log-db.tex:
```diff
@@ -263,6 +311,21 @@
+It creates a temporary trigger that prunes, after an insert, up to
+\var{limit} rows of the specified \var{table} where the \code{erlang:now}
+timestamp in \var{column} is older than \var{max-days}.
+\var{max-days} must be a nonnegative fixnum and
```
add a comma after `fixnum`

laqrix replied:
Updating as suggested.

burgerrg reviewed (commented):
I've done a preliminary review of the documentation but haven't looked at the code yet. Thanks for making these helpful changes!

burgerrg reviewed (approved):
Thank you for this impressive piece of work!

owaddell commented:
Thanks for the review, Bob! The next PR should be smaller.

laqrix commented:
Thanks for the review Bob!
